### ISSUE TITLE
pre_token_generation Lambda injects claims in id_token instead of auth_token

### ISSUE THREAD

**Issue by soulshake** [localstack/localstack#2896]
# Type of request: This is a ...

[x] bug report
[ ] feature request

# Detailed description

I am using a pre-token-generation Lambda trigger to inject claims into JWTs issued by Cognito.

There are two problems with the JWTs issued by LocalStack Cognito:

#### 1. The Lambda seems to be injecting the extra claim into the wrong token.

Tokens issued by our production Cognito user pool have the injected claim in the `id_token`, not in the `access_token`.

With a LocalStack user pool with the same Lambda trigger, it's the other way around: The `access_token` has the extra claim, but the `id_token` does not.

See examples below (the claim in question is `https://hasura.io/jwt/claims`).

#### 2. The token signature can't be validated.

(Tested by pasting the tokens into jwt.io along with the jwks.json key)

## Expected behavior

<details>
<summary>
Example tokens returned by Cognito
</summary>

```
$ aws --profile admin cognito-idp admin-initiate-auth --user-pool-id $POOL_ID  --client-id "$CLIENT_ID" --auth-flow ADMIN_USER_PASSWORD_AUTH --auth-parameters "USERNAME=$USERNAME,PASSWORD=$PASSWORD,SECRET_HASH=$secret_hash"
{
    "ChallengeParameters": {},
    "AuthenticationResult": {
        "AccessToken": "eyJraWQiOiJTVE1QODgxd1hTR0xpNG5XSjllT0xDT2wyWUc3RCtLUHdQTkRjM2FaV2lnPSIsImFsZyI6IlJTMjU2In0.eyJzdWIiOiJlMjhlYmRiNC0xMTIyLTQxMDctYWI1Yy00ZWJiYTc2NWM0YzIiLCJldmVudF9pZCI6ImRlOTRhMDBlLTk0ZWEtNDFhOC04NmU2LWYyYzg2Y2ZmYmQyYSIsInRva2VuX3VzZSI6ImFjY2VzcyIsInNjb3BlIjoiYXdzLmNvZ25pdG8uc2lnbmluLnVzZXIuYWRtaW4iLCJhdXRoX3RpbWUiOjE1OTg2MDMxNjgsImlzcyI6Imh0dHBzOlwvXC9jb2duaXRvLWlkcC51cy1lYXN0LTIuYW1hem9uYXdzLmNvbVwvdXMtZWFzdC0yX1hFT1Nnd0VhayIsImV4cCI6MTU5ODYwNjc2OCwiaWF0IjoxNTk4NjAzMTY4LCJqdGkiOiJiZTdkMDM0Zi1hMjY1LTQ2YzktODE4OS1kMWZkZTEwNzg0MmQiLCJjbGllbnRfaWQiOiI3cnZtZG5rbmw2MnNzMWI3bmVxOWZrdnMybCIsInVzZXJuYW1lIjoiZTI4ZWJkYjQtMTEyMi00MTA3LWFiNWMtNGViYmE3NjVjNGMyIn0.DKxqW-DgkI4EtHwNlfafiNn93W1m3ipCJZMRtWz3M-733sALD1VeFRMyd9h--LR2GHRXgSeRpbiS75tPObm18ltMQUhwMTsAowsW6rhIY_UfmbGH9Ne8xmYO6POLPX_HbZRBT_doJep1HvY4TTFVPd8i4v31MF0yE1Wx9bj-sx82zDddR803bxXnhbSbEos9hLV8FKdINWnFk5YmBROFS4XwJuO0D2oHO69_Vugx1qMDO_TOR8dQ1EeOYiIA9vazYM12G9TcOrL-YxM8WDqIEZlkPlPal6ijFeO0zJI9H4KMZYefi14yn1MnK4MqE2KnWcHX3QDL-oNwlgkejIDcyw",
        "ExpiresIn": 3600,
        "TokenType": "Bearer",
        "RefreshToken": "eyJjdHkiOiJKV1QiLCJlbmMiOiJBMjU2R0NNIiwiYWxnIjoiUlNBLU9BRVAifQ.ftPEVjUE8CpTBlIm8ckDMG5M_Mjwdw_F4FFXYi4u4TBoGFYsoLufQ-CZpGjHGGllgjNLbIRRvS5xS5zy9YD-UeOwRwD-mZOT9DCYJnP_wiUSJqdsmnUyWV9lQLtDEY1R35-0ZoWB9eeQ4ZiRo0wpSI5LN1Kay3CysX31Hnlrruef67b-8Rg0TC_vYYEK8wuludQakpcVFFOefMwYILgDg-1weoc10wnK3wjkaL0O7Zw6MTz-Qp8nuTUml667HRLN3fjcS0w_aMf0MtR6TdOysM3wFxR0Xjl-HlR6n4RQv2MHVY8ud4DkvYHyKpKHgZt2XycYF9EOBPEPtYewoKspyw.DHQEahk7XCRqRWGi.ez3u31x3LJGKQw1tyZr77Ovnnz1nMfAzbvKzXkCFqcd4QlAJDtIApuhRm7509qgeT7Gq3VS-2bN1lBSavwT2NII232CpXFpLOu2JazltmKINNYDsoPidkO8gY6O5i1GxoRa1QXhhiIVHmn68ikjvgcIR8ROYat_32GMJBqRtxWjco1k-d_tC_bHw3jZRCQXRlgzyi2YnwQCrUfjjylzuarw8Tya9-xC9eeHhvcl5-E_In2G3qjzfwZXo2cacL8jV9Sgs_GAzqgMcPnsOZBLUZvAQeB-1Y9aZ48ltZ7yXp7O8-FtfGBgOESnWO1KdgstOdYL9qGo0Um6DDF7E7eDO3ft7hVGjLBY9adC_uw4JXJzgLMPKwPFvn5H84PjFUmsxgQHNDRFdQX9O8wNtg5eesh03bixO9tkQS6WPmenG106Nno4L8U7s7QOSrEd8b4SH-uAabRcUBrX5GegR_9gEVl5EyoyXc3Vnek2kfLJBUZEsIxvj90LkEh_Q7XEvSmdkizDcSgNkyIiCAEQRzDdCnZMqZgpcRo7wxDD8f4lGADC1mY-URhVTgfjHtRUXSMrRR7yWD3_IbYIvDZ6yM8CFfJX7q6wgLajeAGKglpIXWQXMEm81F6ozmbav64v4EfenHHl_6mEJjz0feHMlqnpjx-iY3FqVSKGDqAFPj4kRnBPc3-KIth5S-c35DB4OR1mZbOK1kcOIAdfMJoINBHuEJZJrGRCTsOpWpwS_slUS9FEEI8-cidku7PYTPb7l5iw9aw3MBgo0jAEmJpXDhl-JugQKpfzwIBTrO0KrvBAMf2pkniSels07hh1Mkm6M300759gr28JzjHRCTybkm7jKY3IxIfbHmLpWg58dLSP6jYdH9f1Spcoytyr2Ew1GPIBohtMRDNtOxIsWGe6Mxv2WYYtVZGqVrmhtJptN0tHb2UviVu79CQvO_I3_tOy7EcOwMpUHNb_FyZ8wB9fK3xJrjpSoczm3Y1OBrq38H2Q_XMEMoJlKT8vr0V6T068qwHpyrZwN09TkfdsWDLX_z_rpOj3xl3Up4lyelOHHFYkKRdZajjG8yaj22x3jMh7KPOmlBdWvGifXfqs76mB1_HqOVk70JTLSdEaOYfmcA1F8ZLxLJ5qfpuXu9mQrTIIyur8fuH1BnH6sC2vqp5Iy65UEZ_tJU0hnRdVPeaSBPO9U2Qv689efkN6RLhMDGEcrLa1rzL-swstc4akWTfdBvu-gWN64NJLUU8R8EKZdPqwnDDVeC5gd-OWxqcLGtNayo7htM0EzJ1oE9GB3mERkBQ_VfArcqyHhskU9ymwQ6Sw9asUuq2mqbxecsYckyUk.QAwFJXxvT5WZDm6M_KBstw",
        "IdToken": "eyJraWQiOiJvTjdVQUhkWjVsNTJLSEtMY1JsZjVJeGlRaHN1T1RabEl4Y1wvbCs3RDdMaz0iLCJhbGciOiJSUzI1NiJ9.eyJzdWIiOiJlMjhlYmRiNC0xMTIyLTQxMDctYWI1Yy00ZWJiYTc2NWM0YzIiLCJhdWQiOiI3cnZtZG5rbmw2MnNzMWI3bmVxOWZrdnMybCIsImV2ZW50X2lkIjoiZGU5NGEwMGUtOTRlYS00MWE4LTg2ZTYtZjJjODZjZmZiZDJhIiwidG9rZW5fdXNlIjoiaWQiLCJodHRwczpcL1wvaGFzdXJhLmlvXC9qd3RcL2NsYWltcyI6IntcIlgtSGFzdXJhLUFsbG93ZWQtUm9sZXNcIjpbXCJ1c2VyXCJdLFwiWC1IYXN1cmEtRGVmYXVsdC1Sb2xlXCI6XCJ1c2VyXCIsXCJYLUhhc3VyYS1Vc2VyLUlkXCI6XCJlMjhlYmRiNC0xMTIyLTQxMDctYWI1Yy00ZWJiYTc2NWM0YzJcIn0iLCJhdXRoX3RpbWUiOjE1OTg2MDMxNjgsImlzcyI6Imh0dHBzOlwvXC9jb2duaXRvLWlkcC51cy1lYXN0LTIuYW1hem9uYXdzLmNvbVwvdXMtZWFzdC0yX1hFT1Nnd0VhayIsImNvZ25pdG86dXNlcm5hbWUiOiJlMjhlYmRiNC0xMTIyLTQxMDctYWI1Yy00ZWJiYTc2NWM0YzIiLCJleHAiOjE1OTg2MDY3NjgsImlhdCI6MTU5ODYwMzE2OCwiZW1haWwiOiJkZWxldGVtZUBiZWFjb24uYmlvIn0.E2BN_-Py0P1NlLL1k68IDdd-8udSbf5A9yl6zxjDw5yAFQ6zz7XgHjSZgPOkTJ3ZIa3UhjBNFPah4WCFaJR9S9DMm7kfFFoSTJqsucUJePhIeqarbeTNh-xCAOssupQ4wpojI8hBL1QGVs36XxvGLAAdJM8395w7zGhM3VEqj4i-JrDEfzQOTn_ootOESDwwLOMQwXo0TMOg-QKrcFTUqMFEoBB4sMUGq-_3RlEv7XM4M5nyj60N9wc8m9peTQa3tn9-37jMrf4FYjfoR4FF8IzxgHFKkf27yygEWmCS7fT_O-qvb8_KwjXQv738-xKfcnpPJVjhAozD4LNHY8GSyg"
    }
}
```
</details>

<details>
<summary>
Decoded AccessToken
</summary>

```
HEADER:ALGORITHM & TOKEN TYPE

{
  "kid": "STMP881wXSGLi4nWJ9eOLCOl2YG7D+KPwPNDc3aZWig=",
  "alg": "RS256"
}
PAYLOAD:DATA

{
  "sub": "e28ebdb4-1122-4107-ab5c-4ebba765c4c2",
  "event_id": "de94a00e-94ea-41a8-86e6-f2c86cffbd2a",
  "token_use": "access",
  "scope": "aws.cognito.signin.user.admin",
  "auth_time": 1598603168,
  "iss": "https://cognito-idp.us-east-2.amazonaws.com/us-east-2_XEOSgwEak",
  "exp": 1598606768,
  "iat": 1598603168,
  "jti": "be7d034f-a265-46c9-8189-d1fde107842d",
  "client_id": "abc123etc",
  "username": "e28ebdb4-1122-4107-ab5c-4ebba765c4c2"
}
```
</details>

<details>
<summary>
Decoded IdToken
</summary>

```
HEADER:ALGORITHM & TOKEN TYPE

{
  "cty": "JWT",
  "enc": "A256GCM",
  "alg": "R
{
  "kid": "oN7UAHdZ5l52KHKLcRlf5IxiQhsuOTZlIxc/l+7D7Lk=",
  "alg": "RS256"
}
PAYLOAD:DATA

{
  "sub": "e28ebdb4-1122-4107-ab5c-4ebba765c4c2",
  "aud": "abc123etc",
  "event_id": "de94a00e-94ea-41a8-86e6-f2c86cffbd2a",
  "token_use": "id",
  "https://hasura.io/jwt/claims": "{\"X-Hasura-Allowed-Roles\":[\"user\"],\"X-Hasura-Default-Role\":\"user\",\"X-Hasura-User-Id\":\"e28ebdb4-1122-4107-ab5c-4ebba765c4c2\"}",
  "auth_time": 1598603168,
  "iss": "https://cognito-idp.us-east-2.amazonaws.com/us-east-2_XEOSgwEak",
  "cognito:username": "e28ebdb4-1122-4107-ab5c-4ebba765c4c2",
  "exp": 1598606768,
  "iat": 1598603168,
  "email": "deleteme@beacon.bio"
}
```
</details>

## Actual behavior

<details>
<summary>
Example tokens returned by LocalStack
</summary>

```
508267e4fd15:/opt/code/localstack# awslocal cognito-idp admin-initiate-auth --user-pool-id "$POOL_ID"  --client-id "$CLIENT_ID" --auth-flow ADMIN_USER_PASSWORD_AUTH --auth-parameters "USERNAME=$USERNAME,PASSWORD=$PASSWORD,SECRET_HASH=$secret_hash"
{
    "AuthenticationResult": {
        "AccessToken": "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6InRlc3Qta2V5In0.eyJleHAiOjE1OTg2ODk4NTEsImlzcyI6Imh0dHA6Ly9sb2NhbHN0YWNrOjQ1OTAvdXMtZWFzdC0yXzg5ZDkwMThhMTRlNDRhMDJiYzUzODI5NDQ2NTk5ODM0Iiwic3ViIjoiOTQ1NTBjZmMtODNjMS00NzZjLThhMjgtODEwZjIzNjI2ZTUzIiwiYXV0aF90aW1lIjoxNTk4NjAzNDUxLCJjbGllbnRfaWQiOiI1OThuYWN6NHlsNXJlNTA4bGlsMDYxbXA0bCIsImlhdCI6MTU5ODYwMzQ1MSwiZXZlbnRfaWQiOiI1NDk3OTE3NyIsInRva2VuX3VzZSI6ImFjY2VzcyIsInVzZXJuYW1lIjoiY29uZmlybWVkQGJlYWNvbi5iaW8iLCJodHRwczovL2hhc3VyYS5pby9qd3QvY2xhaW1zIjoie1wiWC1IYXN1cmEtQWxsb3dlZC1Sb2xlc1wiOltcInVzZXJcIl0sXCJYLUhhc3VyYS1EZWZhdWx0LVJvbGVcIjpcInVzZXJcIn0ifQ.Tdy2IIv3bZ1RG7BCqEqf4gxxYlGwTirEgPuUX1Xf9YoNbJa175VIkRZ5k2xM3-edCDXltaQFm3gh3yOiUCg4qajmY9oYLOyIQv3OpLWk-TKeJSOZwFr3es8Mo506pasYx6_7tZzO-tkRfbrU5tpq5_T2xwSMUhTVYxgKLYWX88SuUNBbOH4AKhytepLVWXh62x7viV12Sp9jO_8VsFR5Uioxi5VGnCJwcr8wiB-Xv6EZUqt7jCpuS8PsaCUmwwhn8z8EtzhQD1N-fG1kgZw-Ko5SroPd1uIg-ZN-6w2hz6T197B_xeBWcTMsark46rsjZuQLdXPusysGb2NrmCvBmA",
        "ExpiresIn": 86400,
        "TokenType": "JWT",
        "RefreshToken": "e2a92d02",
        "IdToken": "eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6InRlc3Qta2V5In0.eyJleHAiOjE1OTg2ODk4NTYsImlzcyI6Imh0dHA6Ly9sb2NhbHN0YWNrOjQ1OTAvdXMtZWFzdC0yXzg5ZDkwMThhMTRlNDRhMDJiYzUzODI5NDQ2NTk5ODM0Iiwic3ViIjoiOTQ1NTBjZmMtODNjMS00NzZjLThhMjgtODEwZjIzNjI2ZTUzIiwiYXV0aF90aW1lIjoxNTk4NjAzNDU2LCJjbGllbnRfaWQiOiI1OThuYWN6NHlsNXJlNTA4bGlsMDYxbXA0bCIsImlhdCI6MTU5ODYwMzQ1NiwiZXZlbnRfaWQiOiIwMjk2MzA5MCIsInRva2VuX3VzZSI6ImlkIiwiZW1haWwiOiJjb25maXJtZWRAYmVhY29uLmJpbyIsImF1ZCI6IjU5OG5hY3o0eWw1cmU1MDhsaWwwNjFtcDRsIiwiY29nbml0bzp1c2VybmFtZSI6ImNvbmZpcm1lZEBiZWFjb24uYmlvIn0.Q5nyvhVK36akQ8IGJXOffhJ2Ej95Hl-LUx55ZKX1IZi0YbM1MfgJjeND8FPztf6LW2GwB4xiyGqqQnDfnF7bP0KzX5XUmMvN8l1qpG01VUW1ENCr8jNNw2_C5878XgEmB0_TaVBKzDeQEkgcvQOmekppNX_i0Aqp6VLlRbnBYXUnOFrZTvSx4AInH3zLYlvDCt759aV30sgL_r9yEkyUhBvAPxIEuabUlFBe1z3oaX-1aIrWUBnqqCYdykuyUW1g73DAX6Vq93IXe-nnHrHxMKVv-vZmaGiHlQPs-iklcortZB4VUpA3H1par-Tz7K1TjXVttB9GhQv05pj0HUvimA"
    }
}
```
</details>

<details>
<summary>
Decoded AccessToken
</summary>

```
HEADER:ALGORITHM & TOKEN TYPE

{
  "typ": "JWT",
  "alg": "RS256",
  "kid": "test-key"
}
PAYLOAD:DATA

{
  "exp": 1598689851,
  "iss": "http://localstack:4590/us-east-2_89d9018a14e44a02bc53829446599834",
  "sub": "94550cfc-83c1-476c-8a28-810f23626e53",
  "auth_time": 1598603451,
  "client_id": "598nacz4yl5re508lil061mp4l",
  "iat": 1598603451,
  "event_id": "54979177",
  "token_use": "access",
  "username": "confirmed@beacon.bio",
  "https://hasura.io/jwt/claims": "{\"X-Hasura-Allowed-Roles\":[\"user\"],\"X-Hasura-Default-Role\":\"user\"}"
}
```
</details>

<details>
<summary>
Decoded IdToken
</summary>

```
HEADER:ALGORITHM & TOKEN TYPE

{
  "typ": "JWT",
  "alg": "RS256",
  "kid": "test-key"
}
PAYLOAD:DATA

{
  "exp": 1598689856,
  "iss": "http://localstack:4590/us-east-2_89d9018a14e44a02bc53829446599834",
  "sub": "94550cfc-83c1-476c-8a28-810f23626e53",
  "auth_time": 1598603456,
  "client_id": "598nacz4yl5re508lil061mp4l",
  "iat": 1598603456,
  "event_id": "02963090",
  "token_use": "id",
  "email": "confirmed@beacon.bio",
  "aud": "598nacz4yl5re508lil061mp4l",
  "cognito:username": "confirmed@beacon.bio"
}
```
</details>

#### Secret hash not validated

This is probably not really relevant, but I noticed that if I specify a bogus secret hash in the `admin-initiate-call`, the real Cognito returns an error:

```
An error occurred (NotAuthorizedException) when calling the AdminInitiateAuth operation: Unable to verify secret hash for client abc123etc
```

But LocalStack Cognito doesn't seem to check this.

#### Different `TokenType`

I don't know if it's relevant, but I also notice that the Cognito response has
```
        "TokenType": "Bearer",
```
whereas the LocalStack response has
```
        "TokenType": "JWT",
```

#### Suspiciously short `RefreshToken`

Also not sure if this is relevant, but the `RefreshToken` returned by the real Cognito is much longer than the one returned by LocalStack (compare example tokens above).

# Steps to reproduce

## Command used to start LocalStack

`docker-compose up`

## Client code (AWS SDK code snippet, or sequence of "awslocal" commands)

The reproduction repo at https://github.com/soulshake/localstack-issue-2858 can be used to issue tokens.

**Comment by whummer** [comment]
Thanks for reporting @soulshake - this should be fixed in the `latest` version. Can you please pull the Docker image and give it another try? Please let us know if the problem persists. Thanks for your help!

**Comment by jrevels** [comment]
👋 I work with @soulshake who originally found this bug. I'll leave it to them to fully confirm, but it does appear that the `latest` version of localstack fixes our problem and allows us to remove a workaround for this issue from our local deployment. 

Would love to see this fix (and the fix for localstack/localstack#2858) land in a tagged release that we can pin as local dependency 🙂
